content/drafts/lib-is-for-libraries.md

```diff
@@ -0,0 +1,113 @@
+
+If you're using libraries, you're following the Hollywood Principle: Don't call us, we'll call you. You have control over the application. If you're using a framework, [the framework calls you](https://martinfowler.com/bliki/InversionOfControl.html). In other words, you're just filling in the blanks in a pre-defined structure.
+
+And it's especially pronounced in Rails thanks to the [Active Record pattern](https://www.martinfowler.com/eaaCatalog/activeRecord.html). Active Record is all about mixing data access logic (Rails!) with domain objects (your application!).
+
+## So, what _is_ `lib` for?
+
+In a [2012 post](https://codeclimate.com/blog/what-code-goes-in-the-lib-directory/) on the Code Climate blog, I found this statement:
+
+> I recommend any code that is not specific to the domain of the application goes in `lib/`.
+
+🤔 Now this is quite different. The heart of your application would still live in `app/`, but code that is _not specific to its domain_ would go into `lib/`.
+
+Thinking about code that an application uses that's not specific to its domain, examples that come up for me are HTTP clients, authentication and authorization logic, serialization and deserialization, and so on. For these concerns, we usually don't write the code ourselves - we use libraries, gems. Almost always those are third party gems from [rubygems.org](https://rubygems.org/).
+
+But what if what we need is not available in a well maintained library? What if the functionality we need is so small that it doesn't warrant introducing an external dependency? Something that we can easily write ourselves?
+
+We want the code to live in the same repository as our application, but it's not specific to our application's domain. It's still library code. So it should go into `lib/`.
+
+_It's right there in the name!_
+
+**`lib/` is for libraries.**
+
+## Boundaries
+
+The main advantage of separating library code from your application instead of just lumping it in with all of your other stuff is the reduction of cognitive load. You can look at the library code and understand it without having to know any of the details of your application.
+
+This, more generally, is the advantage we get from having strong architectural boundaries. Robert Martin defines boundaries as lines in an architecture diagram that are only crossed by dependencies in one direction. The application depends on libraries, but libraries never depend on the application.
+
+Sandi Metz, in her book _Practical Object-Oriented Design in Ruby_, explains why this matters:
+
+> "I know who you are and because of that I know what you do." This knowledge is a dependency that raises the cost of change.
+
+Dependencies are _knowledge_. And the less you as a developer need to know to do your work the faster you can get things done, delivering value to your users and checking items off your to-do list.
```

Comment:
I'm struggling to articulate why, but this is the part of this post that feels jankiest to me, even though I agree with all the conclusions.

**Agreed:** 

- the less you need to know about a package's internals, the faster you can use it, and the less painful changes should be
- dependencies are knowledge - and the more granular the knowledge, the more porous the package boundary

**Friction**

I went and double-checked the context for this quote, and it was specifically about relying on `is_a?`, or `responds_to?`, essentially using type switching instead of relying on message passing. In that sense, "Dependencies are knowledge" is a not-quite-logical inversion of the (IMO) intent of the quote "excessively-specific knowledge is an unintended dependency".

Again, I agree with the premise - a dependency is a kind of knowledge, and an effective use of a dependency sets hard rules about how much/what kind of knowledge you, the operator, should have. But there's a mismatch between the conclusion and the supporting quote.

I don't have any alternative suggestions here - maybe something in one of Jim Weirich's [connascence](https://connascence.io/pages/about.html) talks, or dust off an older David Parnas piece on modularity and secrets?